plugins/volume/org.xen.xapi.storage.fsp/sr.py

```diff
@@ -70,6 +80,52 @@ def set_description(self, dbg, sr, new_description):
     def set_name(self, dbg, sr, new_name):
         util.update_sr_metadata(dbg, 'file://' + sr, {'name': new_name})
 
+    def _createvdi(self, sr, name, path, psize):
+        fsp = importlib.import_module("fsp")
+        cb = fsp.Callbacks()
+        description = "Created by sr.scan()"
+        sharable = False
+        vdi_uuid = str(uuid.uuid4())
+        with VolumeContext(cb, sr, 'w') as opq:
+            image_type = ImageFormat.IMAGE_DIRECTORY
+            image_format = ImageFormat.get_format(image_type)
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    volume = db.insert_new_volume(psize, image_type)
+                    db.insert_vdi(name, description, vdi_uuid, volume.id, sharable)
+                    volume_path = cb.volumeGetPath(opq, str(volume.id))
+            os.symlink(path, volume_path)
+        return vdi_uuid
+
+    def _scan(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        sr_dir = meta['uri'] + '/links'
+        log.debug('scanning: {}\n'.format(sr_dir))
+        if os.path.exists(sr_dir):
+            for filename in os.listdir(sr_dir):
+                path = os.path.join(sr_dir, filename)
+                if os.path.exists(path) and os.path.islink(path):
+                    if not path in meta['vdis']:
+                        vdi_uuid = self._createvdi(sr, filename, path, 0)
+                        meta['vdis'][path] = vdi_uuid
+                        util.update_sr_metadata(dbg, 'file://' + sr, meta)
+                        log.debug('New VDI for symlink {}: uuid={}'.format(filename, vdi_uuid))
+
+    def _cleanvdis(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        new_vdis = {}
+        fsp = importlib.import_module("fsp")
+        cb = fsp.Callbacks()
+        for path in meta['vdis']:
+            with VolumeContext(cb, sr, 'w') as opq:
+                with PollLock(opq, 'gl', cb, 0.5):
+                    with cb.db_context(opq) as db:
+                        if os.path.exists(path) and db.get_vdi_by_id(meta['vdis'][path]) is not None:
```

Comment:
By using the db, I shall:
1. Get all custom keys by using `get_all_vdi_custom_keys()`
2. For those that the key is `path_to_link` and the value does not exist, I call `delete_vdi_custom_key()`. If I understand well, when a vdi is deleted, all the `vdi_custom_keys` are deleted from the table. So this clean-up works only when the user intentionally removes a link. I will come back to this use-case later because I am not sure yet if it is a use-case that would like to support.

Reply:
I have removed _clean_vdis() since the list is cleaned during vdi.destroy().